x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java

```diff
@@ -222,43 +206,236 @@ public void testTierNodesPresent() {
         assertTrue(DataTierAllocationDecider.tierNodesPresent("data_content", nodes));
     }
 
+    public void testTierNodesPresentDesiredNodes() {
+        Set<DesiredNode> nodes = Collections.emptySet();
+
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_hot", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_warm", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_cold", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_content", nodes));
+
+        nodes = Set.of(WARM_DESIRED_NODE, CONTENT_DESIRED_NODE);
+
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_hot", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_warm", nodes));
+        assertFalse(DataTierAllocationDecider.tierNodesPresent("data_cold", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_content", nodes));
+
+        nodes = Set.of(DATA_DESIRED_NODE);
+
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_hot", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_warm", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_cold", nodes));
+        assertTrue(DataTierAllocationDecider.tierNodesPresent("data_content", nodes));
+    }
+
     public void testPreferredTierAvailable() {
-        DiscoveryNodes nodes = DiscoveryNodes.builder().build();
+        {
+            final var nodes = DiscoveryNodes.builder().build();
+            final DesiredNodes desiredNodes = null;
```

Comment:
Not sure I follow why we have this variable? Perhaps you intended to randomize it to sometimes have desired nodes?